Migrate groups API to fastAPI

arash77 opened the pull request:
## What did you do?

- Include an Integration test under `test/integration/`
- Refactor the legacy `groups` API logic so all the logic is contained in the `GroupsManager` class
- Create pydantic models for the responses and payloads
- Create the FastAPI routes that will replace the old ones

related to: https://github.com/galaxyproject/galaxy/issues/10889

## Why did you make this change?
This is part of the efforts to migrate the full Galaxy API to FastAPI see #10889.

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

arash77 committed:
pydantic models for the responses and payloads

arash77 committed:
Add FastAPI routes for groups

arash77 committed:
Add integration test for groups

arash77 committed:
Update group schema documentation

arash77 committed:
Add API endpoints for groups

davelopez commented on lib/galaxy/managers/groups.py:
```diff
@@ -51,21 +48,21 @@ def index(self, trans: ProvidesAppContext):
         group = model.Group(name=name)
         sa_session.add(group)
-        encoded_user_ids = payload.get("user_ids", [])
+        encoded_user_ids = getattr(payload, "user_ids", [])
```
Why use `getattr` here? payload seems to be now a Pydantic model.

arash77 replied:
Thanks for the feedback. I've updated the pull request with the modifications.

davelopez commented on lib/galaxy/schema/groups.py:
```diff
@@ -0,0 +1,95 @@
+        title="name of the group",
+        description="name of the group",
+    )
+    user_ids: List[EncodedDatabaseIdField] = Field(
```
For `input` models that contain database IDs you should use `DecodedDatabaseIdField` this will handle the validation and decoding of the ID automatically and you can directly access the decoded ID (int) when using them.

davelopez commented on lib/galaxy/webapps/galaxy/api/groups.py:
```diff
@@ -2,61 +2,85 @@
+        self,
+        trans: ProvidesAppContext = DependsOnTrans,
+    ) -> GroupIndexListResponse:
+        """GET /api/groups - Displays a collection (list) of groups."""
```
These docstrings will appear in the interactive documentation under `http://127.0.0.1:8080/api/docs` once you view the details of that particular route. You should use this to write detailed documentation of the endpoint if needed, but in this case, it is probably fine to just remove it if it's not adding anything to the summary.

The same for the rest of the endpoints.

davelopez replied:
You can completely remove the docstrings line if there is no detailed docs for he endpoint :)

davelopez commented on lib/galaxy/webapps/galaxy/api/groups.py:
```diff
@@ -2,61 +2,85 @@
+        encoded_group_id: EncodedDatabaseIdField = Path(),
+    ) -> GroupShowResponse:
+        """GET /api/groups/{encoded_group_id} - Displays information about a group."""
+        group_id = EncodedDatabaseIdField.decode(encoded_group_id)
```
Use a `DecodedDatabaseIdField` for any input parameter or payload model field, this will validate and auto-decode the ID so you don't need to manually decode it like here.

davelopez commented on test/integration/test_groups.py:
```diff
@@ -0,0 +1,122 @@
+from galaxy_test.driver import integration_util
+
+
+class TestGroupsApi(integration_util.IntegrationTestCase):
```
There seem to be existing API tests for groups in `lib/galaxy_test/api/test_groups.py` and there is no configuration change in Galaxy needed, so these integration tests are unnecessary.

davelopez commented on lib/galaxy/schema/groups.py:
```diff
@@ -0,0 +1,95 @@
+class GroupIndexResponse(Model):
+    """Response schema for a group."""
+
+    model_class: str = Field(
```
The `model_class` here is not just a string, is a particular literal and must always be the same. Check out https://github.com/galaxyproject/galaxy/blob/05fe2868f3a8af1b634579f5c3ca2b47d538e357/lib/galaxy/schema/schema.py#L431 you can also find that there are existing models that you can reuse for groups. You can move them to /galaxy/schema/groups.py` if you want.

davelopez reviewed (commented):
Nice first attempt!

Find a couple of comments below explaining a bit some of our conventions.

arash77 committed:
Refactor group manager and API endpoints

arash77 committed:
Refactor API group endpoints

arash77 committed:
Refactor group manager methods and schema

davelopez commented on lib/galaxy/managers/groups.py:
```diff
@@ -119,25 +112,17 @@ def _get_group(self, sa_session: galaxy_scoped_session, group_id: int) -> model.
+        self, sa_session: galaxy_scoped_session, user_ids: List[DecodedDatabaseIdField]
     ) -> List[model.User]:
-        user_ids = self._decode_ids(encoded_user_ids)
         return get_users_by_ids(sa_session, user_ids)
```
Looks like this function is no longer relevant and can be replaced directly by `get_users_by_ids`

arash77 replied:
Thank you for the review. I will apply them ASAP.

davelopez commented on lib/galaxy/managers/groups.py:
```diff
@@ -119,25 +112,17 @@ def _get_group(self, sa_session: galaxy_scoped_session, group_id: int) -> model.
     ) -> List[model.Role]:
-        role_ids = self._decode_ids(encoded_role_ids)
         stmt = select(Role).where(Role.id.in_(role_ids))
         return sa_session.scalars(stmt).all()
```
Similar to `get_users_by_ids` this function can be moved to `lib/galaxy/managers/roles.py`

davelopez commented on lib/galaxy/schema/groups.py:
```diff
@@ -0,0 +1,80 @@
+
+    model_class: GROUP_MODEL_CLASS = ModelClassField(GROUP_MODEL_CLASS)
+    id: DecodedDatabaseIdField = Field(
+        ...,
```
It is fine to use the `Required` constant instead of the ellipsis (...) for required fields, the reason the ellipsis was used in all the previous models is because the constant was not available in older versions of Pydantic.

davelopez commented on lib/galaxy/schema/groups.py:
```diff
@@ -0,0 +1,80 @@
+    """Response schema for a group."""
+
+    model_class: GROUP_MODEL_CLASS = ModelClassField(GROUP_MODEL_CLASS)
+    id: DecodedDatabaseIdField = Field(
```
```suggestion
    id: EncodedDatabaseIdField = Field(
```

Sorry, I was not completely clear on how to use the Encoded/Decoded ID handling in models. For **response** models we always use the `Encoded` version of the DatabaseIdField, but for **request** (body, query, or path parameters) we use the `Decoded` version. This ensures that we validate incoming IDs to be properly encoded and provide the decoded value directly for use on the endpoints without manual decoding involved and also ensures that the serialized models in the response are encoded properly when they are returned.

So in summary:
- Input -> Decoded
- Output -> Encoded

davelopez commented on lib/galaxy/schema/groups.py:
```diff
@@ -0,0 +1,80 @@
+    __root__: List[GroupIndexResponse]
+
+
+class GroupShowResponse(Model):
```
This model name seems very coupled with the "show" endpoint, but it can be used in other endpoints like the "create".

arash77 replied:
I'm a bit confused. Isn't `GroupShowResponse` exclusively for the 'show' endpoint because of `roles_url` and `users_url`?

davelopez replied:
Sorry I was not very clear, see this comment https://github.com/galaxyproject/galaxy/pull/17051/files#r1400813994

I was just suggesting that maybe `GroupResponse` is a more generic name because `Show` is very tied to one of the endpoints.

davelopez commented on lib/galaxy/schema/groups.py:
```diff
@@ -0,0 +1,80 @@
+    """Response schema for showing a group."""
+
+    model_class: GROUP_MODEL_CLASS = ModelClassField(GROUP_MODEL_CLASS)
+    id: str = Field(
```
```suggestion
    id: EncodedDatabaseIdField = Field(
```

davelopez commented on lib/galaxy/webapps/galaxy/api/groups.py:
```diff
@@ -2,61 +2,79 @@
+        self,
+        trans: ProvidesAppContext = DependsOnTrans,
+        payload: GroupCreatePayload = Body(...),
+    ) -> GroupIndexListResponse:
```
I was referring to this, the create endpoint returns a single Group response so the same response model is used for the `show` and `create` edpoints.

```suggestion
    ) -> GroupShowResponse:
```

arash77 replied:
I used `GroupIndexListResponse` for `create` endpoint because it returns a list with a single group response. but I can make it to not create a list but just a group response.

davelopez replied:
I was not expecting a list here... I wonder if it's a bug I introduced or if it was like that in the original API so we need to keep it for backward compatibility I can't remember it... :sweat_smile:

arash77 replied:
It was a list from the original API:
https://github.com/galaxyproject/galaxy/blob/e8a288b2d389b756c651cbe1a5682239dd6233fe/lib/galaxy/managers/groups.py#L77

davelopez replied:
Yes, I checked back until [the original API endpoint was created in 2012](https://github.com/galaxyproject/galaxy/commit/9bcbabfaa9d1a8a70d4e6e7155015a3a4c8d8c97#diff-f36fa9ebf45ae5ff33bbf8f2749b36b9677ee1fb1600026e61d1c68eabf3059dR70) and it also returns a list with one single group... weird... but I guess it needs to be backward compatible, so all good :+1: 

Sorry for the noise!

arash77 committed:
Refactor group API and role manager
Change create response to a single group response instead of a list
Add response for update
Fixing url parameter

arash77 committed:
Remove unused import statement in groups.py

arash77 committed:
Edit create method to retun list

arash77 committed:
Update GroupResponse to GroupListResponse in API
schema

davelopez commented on lib/galaxy/managers/groups.py:
```diff
@@ -47,33 +35,33 @@ def index(self, trans: ProvidesAppContext):
             item = group.to_dict(value_mapper={"id": DecodedDatabaseIdField.encode})
             encoded_id = DecodedDatabaseIdField.encode(group.id)
-            item["url"] = url_for("group", id=encoded_id)
+            item["url"] = self._url_for(trans, "index", group_id=encoded_id)
```
These URLs don't look correct. It should be returning something like: `/api/groups/{encoded_group_id}` but this one is returning `/api/groups/{encoded_group_id}/roles`. You need to play with the `name` of the FastAPI route to make it generate the route you need.

davelopez commented on lib/galaxy/managers/groups.py:
```diff
@@ -83,32 +71,48 @@ def show(self, trans: ProvidesAppContext, group_id: int):
+        item["roles_url"] = [
+            self._url_for(trans, "group_role", group_id=encoded_id, role_id=group_role)
+            for group_role in encoded_role_id
+        ]
```
The same for these URLs, they should not be lists but rather something like:
`users_url: "/api/groups/{encoded_group_id}/users"`

I also think some of those URLs are currently wrong in the `dev` branch which kind of confirms that nobody is using them... so we should probably deprecate them and remove them at some point.

davelopez reviewed (commented):
Almost there! Pretty good so far!

arash77 committed:
Fix group URLs in GroupsManager and schema
Add name for group_users and group_roles endpoints

davelopez reviewed (approved):
Awesome work!
Congratulations on your first contribution to Galaxy! :rocket:
